cpp/src/shim/libcurl.cpp

```diff
@@ -116,19 +116,43 @@ CURL* CurlHandle::handle() noexcept { return _handle.get(); }
 
 void CurlHandle::perform()
 {
-  // Perform the curl operation and check for errors.
-  CURLcode err = curl_easy_perform(handle());
-  if (err != CURLE_OK) {
-    std::string msg(_errbuf);  // We can do this because we always initialize `_errbuf` as empty.
+  CURLcode err;
+  int http_code   = 0;
+  int retry_count = 0;
+  int max_retries = 5;
+
+  do {
     std::stringstream ss;
-    ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
-    if (msg.empty()) {
-      ss << "(" << curl_easy_strerror(err) << ")";
+
+    CURLcode err = curl_easy_perform(handle());
+    curl_easy_getinfo(handle(), CURLINFO_RESPONSE_CODE, &http_code);
+
+    // Check if we should retry based on HTTP status code
+    if (http_code == 500 || http_code == 502 || http_code == 503 || http_code == 504) {
```

Comment:
Should the list of codes to retry be configurable?

Should 429 be in the list of defaults? https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429 (Too Many Requests). This commonly includes a Retry-After header telling you when to try again.

Reply:
Yes, and possibly 400 Bad Request? (Sometimes the issue is corruption in the request sent from client to server)

Reply:
I think your thoughts are good! I'm happy for us to follow your recommendations here. Thanks for doing that research.